src/Microsoft.Azure.WebJobs.Host/Blobs/Listeners/ScanBlobScanLogHybridPollingStrategy.cs

```diff
@@ -0,0 +1,162 @@
+        private int _scanBlobThreshold = 5000;
+        private readonly IDictionary<IStorageBlobContainer, PollStrategy> _pollStrategies;
+        private readonly IDictionary<IStorageBlobContainer, ICollection<ITriggerExecutor<IStorageBlob>>> _registrations;
+        private ScanContainersStrategy _scanContainerStrategy;
+        private PollLogsStrategy _pollLogStrategy;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private bool _disposed;
+
+        public ScanBlobScanLogHybridPollingStrategy() : base()
+        {
+            _registrations = new Dictionary<IStorageBlobContainer, ICollection<ITriggerExecutor<IStorageBlob>>>(
+                new StorageBlobContainerComparer());
+            _pollStrategies = new Dictionary<IStorageBlobContainer, PollStrategy>(
+                new StorageBlobContainerComparer());
+            _scanContainerStrategy = new ScanContainersStrategy();
+            _pollLogStrategy = new PollLogsStrategy();
+            _cancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public void Start()
+        {
+            ThrowIfDisposed();
+            // can not start the 2 strategies till all regiesterations are done
+            ScnaContainersAndRegisterWithStrategy(_cancellationTokenSource.Token);
+            _scanContainerStrategy.Start();
+            _pollLogStrategy.Start();
+        }
+
+        public void Cancel()
+        {
+            ThrowIfDisposed();
+            _scanContainerStrategy.Cancel();
+            _pollLogStrategy.Cancel();
+            _cancellationTokenSource.Cancel();
+        }
+
+        public Task RegisterAsync(IStorageBlobContainer container, ITriggerExecutor<IStorageBlob> triggerExecutor, CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+            // Register and Execute are not concurrency-safe.
+            // Avoiding calling Register while Execute is running is the caller's responsibility.
+
+            ICollection<ITriggerExecutor<IStorageBlob>> containerRegistrations;
+            if (_registrations.ContainsKey(container))
+            {
+                containerRegistrations = _registrations[container];
+            }
+            else
+            {
+                containerRegistrations = new List<ITriggerExecutor<IStorageBlob>>();
+                _registrations.Add(container, containerRegistrations);
+            }
+
+            containerRegistrations.Add(triggerExecutor);
+            return Task.FromResult(0);
+        }
+
+        public async Task<TaskSeriesCommandResult> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+            await Task.WhenAll(_scanContainerStrategy.ExecuteAsync(cancellationToken),
+                                _pollLogStrategy.ExecuteAsync(cancellationToken));
+
+            // Run subsequent iterations at 2 second intervals.
```

Comment:
what is happening in this approach is, every time I am polling a page out of each container, small containers will always be fully scanned every time, big ones will complete when the budget allows. this approach is the same complexity as the other one in terms of carrying around continuation tokens and paging, but it makes sure that small enough containers are polled every time even if the number of containers increases. while at the same time keeps the load (number of queried blobs) under control